### PR TITLE
Fixes #30105 - Adding reactjs with EmptyState Component

### PR DESCRIPTION
**Pull request by rabajaj0509**
Tasks involved in this effort are as follows:

* [x]  Adding Example Component
* [x]  Adding Foreman metapackages (test, lint, storybook, vendor)
* [x]  Adding API usage via APIMiddleware
* [x]  Use EmptyState from Foreman core
* [x]  work out packaging

### CODE REVIEW

**Comment by rabajaj0509** [comment]
I have few questions:
a) The empty state component does not have any actions to be performed, not sure how to configure the redux infrastructure for this PR.
b) How to move the component to the centre of the page?
c) What is the importance of the this line [here ](https://github.com/theforeman/foreman_discovery/pull/503/files#diff-52f1696d7dae32f43202243e041b3af8R7). To register a plugin we need the `type` parameter in the `componentRegistry.register` function, but not sure how it is important. Does it behave as the entry point?

File: webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js
Comment (sharvit) [inline webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js]: We have a foreman-core component that should give you the same empty-state-pattern, but using pf3:
https://github.com/theforeman/foreman/tree/develop/webpack/assets/javascripts/react_app/components/common/EmptyState

I am wondering if we should migrate our current empty-state-component in foreman-core and reuse it here so all the other pages would enjoy the same pf4 design and it would be consistent over core and plugins.

Comment (sharvit) [reply webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js]: https://foreman-storybook.surge.sh/?path=/docs/components-empty-state-pattern--default-story

Comment (rabajaj0509) [reply webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js]: This sounds interesting, let me try to open a PR in foreman core for this.

**Comment by sharvit** [comment]
> a) The empty state component does not have any actions to be performed, not sure how to configure the redux infrastructure for this PR.

Redux is optional so you don't have to use it. Your `DiscoverdHosts/index.js` can look something like:
```js
import DiscoverdHosts from './DiscoverdHosts';
export default DiscoverdHosts;
```

> b) How to move the component to the centre of the page?

I would check how the core-empy-state achieves that

> c) What is the importance of the this line [here ](https://github.com/theforeman/foreman_discovery/pull/503/files#diff-52f1696d7dae32f43202243e041b3af8R7). To register a plugin we need the `type` parameter in the `componentRegistry.register` function, but not sure how it is important. Does it behave as the entry point?

You don't actually register your plugin (it could be nice if we had that), you register component so you can render them in a .erb file.

So when you call `<%= react_component('DiscoveredHosts') %>`, it would get to here:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/componentRegistry.js#L95-L108

* We don't register all the component, just the one we want to render from a .erb file

File: package.json
Comment (Ron-Lavi) [inline package.json]: can you upgrade it to use "^4.11.0" which is the current version?

File: webpack/src/ForemanDiscovery/DiscoveredHosts/index.js
Comment (Ron-Lavi) [inline webpack/src/ForemanDiscovery/DiscoveredHosts/index.js]: I agree with Avi, for this initial PR, let's not use redux yet since we don't use the action/reducers or some redux logic.

File: webpack/src/ForemanDiscovery/DiscoveredHosts/DiscoveredHosts.js
Comment (Ron-Lavi) [inline webpack/src/ForemanDiscovery/DiscoveredHosts/DiscoveredHosts.js]:
```suggestion
export { default } from './Components/EmptyState';
```

**Review by Ron-Lavi — COMMENTED** [review]
Nice work! tested and it works well :+1: 
- I would get rid of the redux files, e.g: actions/reducers and add them when we need in a follow-up.
- it would be awesome if we can use the EmptyState from Foreman and contribute the PF4 implementation.
- noticed some linting issues when running `npm run lint`
- can you add a simple snapshot test to the component?
- want to add a github action to run the JS tests? something like [REX workflow](https://github.com/theforeman/foreman_remote_execution/blob/master/.github/workflows/ci.yml#L75-L101)

File: config/routes.rb
Comment (rabajaj0509) [inline config/routes.rb]: @lzap can you please check if there is no discovered host, the welcome page is displayed and when there is one discovered host we see the table in place?

Comment (rabajaj0509) [reply config/routes.rb]: I checked and it fails :( @laviro @sharvit any idea how i can  just show the welcome page when there is no discovered host present. When a discovered host is present, we need to show the discovered host table.

Comment (Ron-Lavi) [reply config/routes.rb]: as discussed offline, lets let the rails controller decide on the view for this PR,
next step would be to let React decide what to show - empty state when there's no hosts / discovered hosts page.

File: app/controllers/discovered_hosts_controller.rb
Comment (rabajaj0509) [inline app/controllers/discovered_hosts_controller.rb]: @lzap @ezr-ondrej @xprazak2  I am trying to redirect to the `foreman_discovery/react` controller I have written for react but it doesn't seem to redirect. Can you take a look please?

**Comment by lzap** [comment]
Please add to the TODO list:

* work out packaging
* test on production install (installed from package)

Thanks

File: app/controllers/discovered_hosts_controller.rb
Comment (xprazak2) [reply app/controllers/discovered_hosts_controller.rb]: There is no route for react controller action in the config/routes, it will redirect if you add something like:

```rb
get 'discovered_hosts/empty', to: 'foreman_discovery/react#index'
```

That will redirect you to a completely new page (with different URL) that shows the empty state:



I think you want to stay on the same page, just displaying the empty component. You can do that by mounting the EmptyState component directly into the welcome.html.erb

File: app/views/foreman_discovery/layouts/application_react.html.erb
Comment (xprazak2) [inline app/views/foreman_discovery/layouts/application_react.html.erb:11]: No need to bind data into the attributes, current user and taxonomies should be available in store.

File: app/controllers/discovered_hosts_controller.rb
Comment (rabajaj0509) [reply app/controllers/discovered_hosts_controller.rb]: Awesome! thanks @xprazak2  :) It works like a charm!

File: package.json
Comment (tbrisker) [inline package.json]: typo:
```suggestion
    "url": "http://projects.theforeman.org/projects/foreman_discovery/issues"
```

File: package.json
Comment (tbrisker) [inline package.json]: this should match the plugin version

File: package.json
Comment (tbrisker) [inline package.json]: missing

File: package.json
Comment (tbrisker) [inline package.json]: this should probably be higher

File: package.json
Comment (tbrisker) [inline package.json]: is this actually needed? it is in foreman core

Comment (Ron-Lavi) [reply package.json]: yeah I don't think we need it, and at least for now we don't use it.. :)

File: app/views/discovered_hosts/welcome.html.erb
Comment (Ron-Lavi) [inline app/views/discovered_hosts/welcome.html.erb:14]: Maybe soon enough https://github.com/theforeman/foreman/pull/6936 will get merged and you could use the router registry instead :)

File: package.json
Comment (Ron-Lavi) [reply package.json]: and all other `@theforeman/*` packages should match that version

File: webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js
Comment (Ron-Lavi) [inline webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js]: should it be split into multi-lines? also.. can we pass a component to the `ForemanEmptyState`?
can you add translations?

File: webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js
Comment (Ron-Lavi) [inline webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js]: can you pass the `documentation` object fixed from the server?

File: webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js
Comment (Ron-Lavi) [inline webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js:28]: Can you add `PropTypes` ?

**Comment by Ron-Lavi** [comment]
@tbrisker @sharvit @lzap regarding packaging, other than adding webpack/ , package.json etc.. to the gemspec file, do we need to do something special?
Atm we don't add any new npm package that doesn't exist in core/ theforeman/vendor

**Comment by xprazak2** [comment]
I do not think you will need anything special, adding the new files and [BuildRequires](https://github.com/theforeman/foreman-packaging/blob/rpm/2.2/packages/plugins/rubygem-foreman-tasks/rubygem-foreman-tasks.spec#L49) entries should do the trick.

**Comment by lzap** [comment]
I don't see anything packaging-related other than JS.

File: app/views/discovered_hosts/welcome.html.erb
Comment (rabajaj0509) [reply app/views/discovered_hosts/welcome.html.erb:14]: Sure, I am also waiting for the core empty state PR to get merged! Will update this PR soon

File: webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js
Comment (rabajaj0509) [reply webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js]: I think we should not split here, rather must do a split in the core PR. What do you think?

File: webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js
Comment (rabajaj0509) [reply webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js]: I am not sure, if I follow this comment. Can you please help with it?

Comment (Ron-Lavi) [reply webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js]: Can you rename the props from `documentation` to `docUrl`?

**Comment by rabajaj0509** [comment]
@lzap, after speaking to @xprazak2 about the packaging, he suggested that, we need to first merge this PR (I am waiting for the Foreman core to be merged first :smiley:) after which a gem release needs to be done and then the packaging PR will take place.

So once the core PR is merge, we can merge this one followed by the packaging PR. Sounds good?

Just to make sure process is quick, following are the changes I intent to make in the packaging PR (not sure if anythign would be required in the changelog): 

```.spec
--- a/packages/plugins/rubygem-foreman_discovery/rubygem-foreman_discovery.spec
+++ b/packages/plugins/rubygem-foreman_discovery/rubygem-foreman_discovery.spec
@@ -33,6 +33,10 @@ BuildRequires: foreman-plugin >= %{foreman_min_version}
 BuildRequires: %{?scl_prefix_ruby}ruby(release)
 BuildRequires: %{?scl_prefix_ruby}ruby
 BuildRequires: %{?scl_prefix_ruby}rubygems-devel
+BuildRequires: %{?scl_prefix}npm(@babel/core) >= 7.7.0
+BuildRequires: %{?scl_prefix}npm(@babel/core) < 8.0.0
+BuildRequires: %{?scl_prefix}npm(@theforeman/builder) >= 4.0.2
+BuildRequires: %{?scl_prefix}npm(@theforeman/builder) < 5.0.0
 BuildArch: noarch
 Provides: %{?scl_prefix}rubygem(%{gem_name}) = %{version}
 Provides: foreman-plugin-%{plugin_name} = %{version}
@@ -94,10 +98,17 @@ cp -pa .%{gem_dir}/* \
 %{gem_libdir}
 %{gem_instdir}/locale
 %exclude %{gem_cache}
+%exclude %{gem_instdir}/.eslintrc
+%exclude %{gem_instdir}/.stylelintrc
+%exclude %{gem_instdir}/package.json
+%exclude %{gem_instdir}/webpack
 %{gem_spec}
 %{foreman_bundlerd_plugin}
 %{foreman_apipie_cache_foreman}
 %{foreman_apipie_cache_plugin}
+%{foreman_assets_plugin}
+%{foreman_webpack_plugin}
+%{foreman_webpack_foreman}

```

**Comment by lzap** [comment]
I could not find a link for the mentioned core PR you need, can you share?

Also make tests passing. Other than that, good work picking a safe component.

**Comment by rabajaj0509** [comment]
> I could not find a link for the mentioned core PR you need, can you share?

Sure, https://github.com/theforeman/foreman/pull/7825, there is a small nit which @laviro @sharvit  can help me with, then that is done and here everything should be green :smile: 

> 
> Also make tests passing. Other than that, good work picking a safe component.

Definitely, will make the tests green and ping you once it is done. It is a safe component indeed :stuck_out_tongue:

**Comment by rabajaj0509** [comment]
@laviro @sharvit  can you provide final approval for the PR so that @lzap can merge this one :+1:

File: app/views/discovered_hosts/welcome.html.erb
Comment (Ron-Lavi) [inline app/views/discovered_hosts/welcome.html.erb:15]: documentation page doesn't seem to exit yet for this version, any chance we could link to a `latest` url?
cc @lzap

File: babel.config.js
Comment (Ron-Lavi) [inline babel.config.js]: I guess it is the same, though usually in other plugins that's the way did it:
```suggestion
module.exports = {
  presets: ['@theforeman/builder/babel'],
};
```

File: webpack/src/reducers.js
Comment (Ron-Lavi) [inline webpack/src/reducers.js]: Currently the console show the following error:
```console
Store does not have a valid reducer. Make sure the argument passed to combineReducers is an object whose values are reducers.
```

For now because you are not passing any reducer it would be OK to return an empty object:
```suggestion
  foremanDiscovery: {},
```

File: webpack/src/ForemanDiscovery/DiscoveredHosts/index.js
Comment (Ron-Lavi) [inline webpack/src/ForemanDiscovery/DiscoveredHosts/index.js]:
```suggestion
const DiscoveredHosts = ({ docUrl }) => <EmptyState docUrl={docUrl} />;
```

**Review by Ron-Lavi — COMMENTED** [review]
Tested and it looks great, left some inline comments :)

File: app/views/discovered_hosts/welcome.html.erb
Comment (rabajaj0509) [reply app/views/discovered_hosts/welcome.html.erb:15]: IMO, the documentation link will be generated only when a new major version of the plugin is released, So the current URL needs no change. Correct me if I am wrong @lzap

File: webpack/index.js
Comment (Ron-Lavi) [inline webpack/index.js:11]: non-blocking as we don't really use redux yet,
but I didn't see `foremanDiscovery` in the redux store.

**Review by Ron-Lavi — APPROVED** [review]
Thanks @rabajaj0509 !!
LGTM, one thing is that the reducer namespace wasn't added to the store,
we don't use it yet so I will not block on it..

about the documentation link address I think we could also discuss it on a follow up PR, and I guess soon after the release there will be a suitable address in foreman.org.

great work :1st_place_medal:

File: app/views/discovered_hosts/welcome.html.erb
Comment (lzap) [reply app/views/discovered_hosts/welcome.html.erb:15]: Yeah it is this status quo mode - it will be only valid after we do the release and new docs. I'd waive this in at this point. I would love to fix all our docs links in Foreman core and create some unified landing page ... someday.

**Comment by lzap** [comment]
Guys, I really appreciate the effort, but this is really bad time for merging this - we are just about to branch off the next release. Unless there is commitment from you guys to sort out all the issues - I can imagine problems with packages both during build and runtime.

**Comment by rabajaj0509** [comment]
> Guys, I really appreciate the effort, but this is really bad time for merging this - we are just about to branch off the next release. Unless there is commitment from you guys to sort out all the issues - I can imagine problems with packages both during build and runtime.

Sure, no issues at all, once the near release it done we can have this for the next cycle. Till then we can focus on the discovery rules page, I remember we also have a rule priority related issue open, which we can solve with the upcoming react PR for it :) 

What do you think? @lzap @laviro

**Comment by lzap** [comment]
I think the patch is not so big and there is zero development in discovery so zero chance getting conflicts, I'd wait for one or two weeks until we branch off. Then we can merge and we have plenty of time to fix things. :-)

**Comment by Ron-Lavi** [comment]
sounds good, thanks all!

**Comment by lzap** [comment]
You know what, let's merge this and test packaging part for 2.2. @rabajaj0509 would you mind checking out tomorrow nightly installation? There is still a week available to fix things. Thanks all.

**Comment by rabajaj0509** [comment]
> You know what, let's merge this and test packaging part for 2.2. @rabajaj0509 would you mind checking out tomorrow nightly installation? There is still a week available to fix things. Thanks all.

Thanks for merging Lukas :) 

Hey, I thought we need to create a packaging PR for this PR immediately after this is merged. I am not good with packaging, so by `lets test packaging part for 2.2` means we don't need the packaging PR for now right? we just need to test tomorrow's nightly and see nothing breaks. Am I correct @lzap ?

**Comment by lzap** [comment]
Which files should be added to the gem?

```
 .eslintrc                                                                        |  7 +++++++
 .gitignore                                                                       |  5 +++++
 .stylelintrc                                                                     |  5 +++++
 app/views/discovered_hosts/welcome.html.erb                                      | 24 +++++++++++++++---------
 babel.config.js                                                                  |  3 +++
 lib/foreman_discovery/engine.rb                                                  |  3 ++-
 package.json                                                                     | 39 +++++++++++++++++++++++++++++++++++++++
 test/functional/discovered_hosts_controller_test.rb                              |  2 +-
 webpack/index.js                                                                 | 18 ++++++++++++++++++
 webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js | 28 ++++++++++++++++++++++++++++
 webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/index.js      |  1 +
 webpack/src/ForemanDiscovery/DiscoveredHosts/index.js                            |  6 ++++++
 webpack/src/reducers.js                                                          |  7 +++++++
```

I *think* it should be: `babel.config.js`, `package.json` and `webpack/` but please confirm. These will need to be updated in SPEC as well.

**Comment by rabajaj0509** [comment]
> Which files should be added to the gem?
> 
> ```
>  .eslintrc                                                                        |  7 +++++++
>  .gitignore                                                                       |  5 +++++
>  .stylelintrc                                                                     |  5 +++++
>  app/views/discovered_hosts/welcome.html.erb                                      | 24 +++++++++++++++---------
>  babel.config.js                                                                  |  3 +++
>  lib/foreman_discovery/engine.rb                                                  |  3 ++-
>  package.json                                                                     | 39 +++++++++++++++++++++++++++++++++++++++
>  test/functional/discovered_hosts_controller_test.rb                              |  2 +-
>  webpack/index.js                                                                 | 18 ++++++++++++++++++
>  webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js | 28 ++++++++++++++++++++++++++++
>  webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/index.js      |  1 +
>  webpack/src/ForemanDiscovery/DiscoveredHosts/index.js                            |  6 ++++++
>  webpack/src/reducers.js                                                          |  7 +++++++
> ```
> 
> I _think_ it should be: `babel.config.js`, `package.json` and `webpack/` but please confirm. These will need to be updated in SPEC as well.

I think `package.json` comes from the foreman core, right @laviro @xprazak2 ?
Also we need to:
```
%exclude %{gem_instdir}/.eslintrc
%exclude %{gem_instdir}/.stylelintrc
%exclude %{gem_instdir}/package.json
%exclude %{gem_instdir}/webpack
```

We require these files for sure:
 
 ```
  app/views/discovered_hosts/welcome.html.erb                                      | 24 +++++++++++++++---------
  lib/foreman_discovery/engine.rb                                                  |  3 ++-
  test/functional/discovered_hosts_controller_test.rb                              |  2 +-
  webpack/index.js                                                                 | 18 ++++++++++++++++++
  webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/EmptyState.js | 28 
  webpack/src/ForemanDiscovery/DiscoveredHosts/Components/EmptyState/index.js      |  1 +
  webpack/src/ForemanDiscovery/DiscoveredHosts/index.js                            |  6 ++++++
  webpack/src/reducers.js                                                          |  7 +++++++
 ```

**Comment by lzap** [comment]
You only need to exclude files which are distributed in the gem. I don't think we need to distribute lint configuration at all. Only those files which are required during runtime or to build the GEM (e.g. extension C files - not this case).

**Comment by lzap** [comment]
foreman_discovery-16.3.1.gem it is!

**Pull request merged**
